Text Selection in Safari: Try new fix for recent version

jasmussen opened the pull request:
## What?

Followup to #43313. A very recent update to the Safari browser has reintroduced the original issue of full-width selection indicators appearing between blocks. This PR dives in and applies a new fix. 

In trunk, selecting text looks like this:
![safari in trunk](https://user-images.githubusercontent.com/1204802/190126231-d92709bc-5d2e-4f31-9beb-65f0e376fae6.gif)

With this PR applied, the selection looks better and more correct in terms of what you've actually selected:

![safari after](https://user-images.githubusercontent.com/1204802/190126316-2b00cd3f-7ec7-4fbf-a48d-351ec90689c1.gif)

## How?

It seems the Safari selection indicator has become sensitive to the presence of `position: relative;` for the purposes of drawing the indicator. For context, that property is set so that we can draw the blue focus rectangle around the selected block in a pseudo element. 

But if set initially, Safari will draw an incorrect selection indicator. It is unfortunately not enough to apply `position: static;` or `unset` or anything like that, the property must simply not be set at all. This is a little curious, and if you have any insights as to why this might be, I'd appreciate it.

Nevertheless, setting the value contextually as this PR does, appears to still allow us to paint the pseudo elements as we need to, without breaking the selection.

## Testing Instructions

* Use a theme with a centered main column, such as TT2 or TT3
* Use Safari, ideally the latest version (17614.1.25.9.10 or newer)
* Select across paragraphs and headings in the editor

The selection marker should look correct.

jasmussen committed:
Text Selection in Safari: Try new fix for recent version

mcsf commented:
👋 Safari is still being difficult with me. I'm on 17614.1.25.9.10 too, with the TT2 theme, and this is what I get:

https://user-images.githubusercontent.com/150562/190141065-080d587a-3409-4e7e-9960-db7d0539a529.mov

* Weird highlight line appears just as I'm selecting the first line of the next block
* Intermittent whole-block selection as the cursor keeps travelling in between blocks
* Sudden highlighting to the left and right of the blocks as the last block is selected

jasmussen commented:
Thank you for testing. One issue that would be excellent to fix, but is beyond the scope to fix, is that when your cursor is _between two blocks_, the selection is treated as a multi-selection (invoking the blue borders) even if it should be counted as a partial selection. If we can fix that so multi selection truly only invokes if a non-text block is selected, that would fix the issue right up until what happens at the 6 second mark, which I'll try my best to reproduce and fix now 😅

jasmussen commented:
While I'm fixing this branch, can you take a quick look at #44150 which is a work in progress refactor?

jasmussen committed:
Fix issue.

jasmussen commented:
Alright @mcsf, I think I've fixed the issue you saw. It was quite hard to reproduce, it helps when the last block you select is the last block on the page it seems, and the state you managed to get it in was one of _multi-selection_, but not _partial selection_. I'll get back to this in a minute, but for now, I think I've fixed the issue you were seeing, can you please test again?

![small fix](https://user-images.githubusercontent.com/1204802/190150636-bcdac3cf-7d48-472c-8bcf-47f833e80a71.gif)

One thing you'll observe here is the one I mentioned: when the cursor is between two blocks, the blue borders appear. Here's what happens, technically:

![partially selected disappears](https://user-images.githubusercontent.com/1204802/190150747-ade7c530-2c08-4eb8-809a-04844b7747a8.gif)

When you are making a partial selection, `is-partially-selected` and `is-multi-selected` classes are both applied to the block. However when your cursor is between two blocks, `is-partially-selected` is briefly removed. If we are able to fix that, so `is-partially-selected` remains with stability right up until the point where we're sure it's becoming a standard multi-selection, the flickering issue you're seeing should also be fixed.

jasmussen commented:
Working on this, noting here some observations on how Safari is drawing selection markers:

* Every element across the markup is selected and highlighted, including pseudo elements. Those can be hidden from the selection using `user-select: none;`, or `height: 0;`.
* Selections are stacked, so the containing element might surface a full-width selection indicator, even if it's an element inside that you are selecting. This stacked selection can be hidden, for example using `.block-editor-block-list__layout::selection { background: transparent; }`.
* Selections are shown as full-width, including any margins applied, unless the selected element is `position: static;` _when the selection is made_. `position: relative;` can be applied without affecting the selection after it has been made.
* The multiple contenteditables in sequence appears to "crop" the selection indication. This explains why in some cases there's a thin selection indicator shown above a paragraph — it is still part of the preceeding paragraph, just cropped.
* A combination of all three above techniques are necessary in order to control the selection indication.

jasmussen commented:
I extracted the flickering issue into a separate issue: #44153.

mcsf commented:
Thanks for the excellent follow-up! Unfortunately I am still getting the wide background selection.

I find it interesting that you found it hard to reproduce, because in my setup/post the result is guaranteed. I don't know what it is, but I'll share my whole content, in case that helps:

<details><pre><code>&lt;!-- wp:paragraph -->
&lt;p>aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa&lt;/p>
&lt;!-- /wp:paragraph -->

&lt;!-- wp:paragraph -->
&lt;p>cccc&lt;/p>
&lt;!-- /wp:paragraph -->

&lt;!-- wp:paragraph -->
&lt;p>dddd&lt;/p>
&lt;!-- /wp:paragraph -->

&lt;!-- wp:paragraph -->
&lt;p>eeee&lt;/p>
&lt;!-- /wp:paragraph -->

&lt;!-- wp:paragraph -->
&lt;p>bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb&lt;/p>
&lt;!-- /wp:paragraph -->
</code></pre></details>

jasmussen commented:
Indeed, I can reproduce with that particular test content, which blows my mind 😅 — thank you, I'll dive in and see what's up!

jasmussen committed:
Better fix.

jasmussen commented:
Alright, I think I have a pretty solid fix:

![status](https://user-images.githubusercontent.com/1204802/190379645-195a0ffd-e66a-497e-913a-2792a26d913a.gif)

Some new revelations:

* Really any stacking layer, unless its `::selection` is made transparent, can almost at random, show up as part of the selection. Thankfully the transparent color hides that. So anything that should never be selected, we can easily apply that to.
* When an element is `position: relative;`, Safari's selection model changes from a _full-width_ version that goes edge to edge of any margins on the element, to a more "inline" version, that only marks the content. In a previous version of this PR, we applied position-relative contextually, but so long as it has to become relative at some point in the future, the value probably has to stay or the switch from _inline_ to _full width_ will be jarring.
* The pseudo element that exists after each block also shows up in the selection indication. Because it is absolutely positioned, it is subject to sub-pixel rounding issues, meaning the selection indication (which has a little built in padding) can occasionally "peek through". The latest commit contextually repositions this pseudo element by half a pixel, so this selection gets literally hidden under the block itself. 

I think this PR is pretty solid, if it fixes the issue for you as well Miguel, I think we should land it.

But an important note, the flickering seen when hovering between two blocks is entirely fixed by https://github.com/WordPress/gutenberg/pull/44154. So we should land that one too!

jasmussen commented:
Also of note: the issue of the pseudo element sub-pixel glitches mentioned can potentially be fixed more or less entirely by #44150, which at the moment tries to refactor that pseudo element away entirely.

jasmussen commented:
Edit, posted in the wrong place, meant to post on #44150.

draganescu reviewed (approved):
I tested and this works excellent! Given the code is pretty specific let's 🚢 it!

ellatrix commented:
Seems fine to me. I'll look into the other issues.

jasmussen commented:
Thanks, I'll land this and rebase the others!

ockham commented:
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: 6082384656

mcsf commented:
@jasmussen: this looks great now! Thank you to everyone for the fixes and reviews.
